cpp/oneapi/dal/algo/triangle_counting/backend/cpu/relabel_kernels.hpp

```diff
@@ -0,0 +1,118 @@
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+
+#include <memory>
+
+#include "oneapi/dal/algo/triangle_counting/common.hpp"
+#include "oneapi/dal/algo/triangle_counting/vertex_ranking_types.hpp"
+#include "oneapi/dal/backend/dispatcher.hpp"
+#include "oneapi/dal/backend/interop/common.hpp"
+#include "oneapi/dal/detail/threading.hpp"
+#include "oneapi/dal/common.hpp"
+#include "oneapi/dal/detail/policy.hpp"
+
+namespace oneapi::dal::preview {
+namespace triangle_counting {
+namespace detail {
+
+template <typename Cpu>
+void sort_ids_by_degree_cpu(const std::int32_t* degrees,
+                            std::pair<std::int32_t, std::size_t>* degree_id_pairs,
+                            std::int64_t vertex_count) {
+    dal::detail::threader_for(vertex_count, vertex_count, [&](std::int32_t n) {
+        degree_id_pairs[n] = std::make_pair(degrees[n], (size_t)n);
+    });
+    dal::detail::parallel_sort(degree_id_pairs, degree_id_pairs + vertex_count);
+}
+
+template <typename Cpu>
+void fill_new_degrees_and_ids_cpu(std::pair<std::int32_t, std::size_t>* degree_id_pairs,
+                                  std::int32_t* new_ids,
+                                  std::int32_t* degrees_relabel,
+                                  std::int64_t vertex_count) {
+    dal::detail::threader_for(vertex_count, vertex_count, [&](std::int32_t n) {
+        degrees_relabel[n] = degree_id_pairs[n].first;
+        new_ids[degree_id_pairs[n].second] = n;
+    });
+}
+
+template <typename Cpu>
+void parallel_prefix_sum_cpu(std::int32_t* degrees_relabel,
+                             std::int64_t* offsets,
+                             std::int64_t* part_prefix,
+                             std::int64_t* local_sums,
+                             size_t block_size,
+                             std::int64_t num_blocks,
+                             std::int64_t vertex_count) {
+    dal::detail::threader_for(num_blocks, num_blocks, [&](std::int64_t block) {
+        std::int64_t local_sum = 0;
+        std::int64_t block_end =
+            std::min((std::int64_t)((block + 1) * block_size), (std::int64_t)vertex_count);
+        for (std::int64_t i = block * block_size; i < block_end; i++) {
+            local_sum += degrees_relabel[i];
+        }
+        local_sums[block] = local_sum;
+    });
+
+    std::int64_t total = 0;
+    for (size_t block = 0; block < num_blocks; block++) {
```

Comment:
why size, not int64?
does the compiler vectorize the loop?

Reply:
changed for int64 and added pragma vector always

Reply:
Mistype, fixed

Reply:
Wasn't vectorized by default, added pragma